src/bsoncxx/lib/bsoncxx/v1/element/view.cpp

```diff
@@ -12,19 +12,337 @@
+    }
+
+    std::uint32_t offset() const {
+        return _offset;
+    }
+
+    std::uint32_t keylen() const {
+        return _keylen;
+    }
+
+    bool is_valid() const {
+        return _raw && _is_valid;
+    }
+
+    impl to_invalid() const {
+        auto ret = *this;
+        ret._is_valid = false;
+        return ret;
+    }
+
+    v1::types::id type_id() const {
+        auto const iter = this->iter();
+        return static_cast<v1::types::id>(bson_iter_type(&iter));
+    }
+
+    v1::types::id type_id_unchecked() const {
+        if (auto const iter_opt = this->iter_unchecked()) {
+            return static_cast<v1::types::id>(bson_iter_type(&*iter_opt));
+        }
+        return v1::types::id{}; // BSON_TYPE_EOD
+    }
+
+    v1::stdx::optional<bson_iter_t> iter_unchecked() const {
+        bson_iter_t iter;
+        if (bson_iter_init_from_data_at_offset(&iter, _raw, _length, _offset, _keylen)) {
+            return iter;
+        }
+        return v1::stdx::nullopt;
+    }
+
+    bson_iter_t iter() const {
+        this->check();
+        if (auto iter_opt = this->iter_unchecked()) {
+            return *iter_opt;
+        }
+        throw v1::exception{code::invalid_data};
+    }
+
+    v1::types::view type_view() const {
+        this->check();
+        return this->type_view_unchecked();
+    }
+
+    v1::types::view type_view_unchecked() const {
+        if (auto opt = v1::types::view::internal::make(_raw, _length, _offset, _keylen)) {
+            return *opt;
+        }
+        throw v1::exception{code::invalid_data};
+    }
+
+    v1::types::value type_value() const {
+        this->check();
+        if (auto opt = v1::types::value::internal::make(_raw, _length, _offset, _keylen)) {
+            return *opt;
+        }
+        throw v1::exception{code::invalid_data};
+    }
+};
+
+void view::impl::check() const {
+    static_assert(is_semiregular<impl>::value, "bsoncxx::v1::element::view::impl must be semiregular");
+    static_assert(is_semitrivial<impl>::value, "bsoncxx::v1::element::view::impl must be semitrivial");
+
+    static_assert(sizeof(view::_storage) >= sizeof(view::impl), "insufficient size");
+    static_assert(alignof(view) >= alignof(view::impl), "insufficient alignment");
+
+    if (!this->is_valid()) {
+        bson_iter_t iter;
+        if (_raw && bson_iter_init_from_data_at_offset(&iter, _raw, _length, _offset, _keylen)) {
+            std::string msg;
+            msg += "last known element key \"";
+            msg += bson_iter_key(&iter);
+            msg += '"';
+            throw v1::exception{code::invalid_view, msg};
+        } else {
+            throw v1::exception{code::invalid_view};
+        }
+    }
+}
+
+view::~view() = default;
+
+view::view(view const& other) noexcept {
+    new (internal::impl(this)) impl{internal::impl(other)};
+}
+
+view& view::operator=(view const& other) noexcept {
+    *internal::impl(this) = internal::impl(other);
+    return *this;
+}
```

Comment:
Ownership semantics are implemented in terms of the `impl` class. As noted in the prior PR and by the static assertions above, the `impl` class is semitrivial (hence straightforward copies) even though `view` itself is not (due to the opaque `impl` class).